tests/test_execution_compute_plan.py

```diff
@@ -0,0 +1,238 @@
+    ]
+
+    # check all traintuples are done and check they have been executed on the expected
+    # node
+    for t in traintuples:
+        assert t.status == 'done'
+
+    traintuple_1, traintuple_2, traintuple_3 = traintuples
+
+    assert len(traintuple_3.in_models) == 2
+
+    assert traintuple_1.dataset.worker == session_1.node_id
+    assert traintuple_2.dataset.worker == session_2.node_id
+    assert traintuple_3.dataset.worker == session_1.node_id
+
+
+def test_compute_plan_single_session_success(factory, session):
+    """A compute plan with 3 traintuples and 3 associated testtuples"""
+
+    # Create a compute plan with 3 steps:
+    #
+    # 1. traintuple + testtuple
+    # 2. traintuple + testtuple
+    # 3. traintuple + testtuple
+
+    spec = factory.create_dataset()
+    dataset = session.add_dataset(spec)
+
+    spec = factory.create_algo()
+    algo = session.add_algo(spec)
+
+    spec = factory.create_data_sample(test_only=False, datasets=[dataset])
+    data_sample_1 = session.add_data_sample(spec)
+
+    spec = factory.create_data_sample(test_only=False, datasets=[dataset])
+    data_sample_2 = session.add_data_sample(spec)
+
+    spec = factory.create_data_sample(test_only=False, datasets=[dataset])
+    data_sample_3 = session.add_data_sample(spec)
+
+    spec = factory.create_data_sample(test_only=True, datasets=[dataset])
+    test_data_sample_1 = session.add_data_sample(spec)
+
+    spec = factory.create_objective(
+        dataset=dataset,
+        data_samples=[test_data_sample_1],
+    )
+    objective = session.add_objective(spec)
+
+    cp_spec = factory.create_compute_plan(algo=algo, objective=objective)
+
+    traintuple_spec_1 = cp_spec.add_traintuple(
+        dataset=dataset,
+        data_samples=[data_sample_1]
+    )
+    testtuple_spec_1 = cp_spec.add_testtuple(traintuple_spec_1)
+
+    traintuple_spec_2 = cp_spec.add_traintuple(
+        dataset=dataset,
+        data_samples=[data_sample_2],
+        traintuple_specs=[traintuple_spec_1]
+    )
+    testtuple_spec_2 = cp_spec.add_testtuple(traintuple_spec_2)
+
+    traintuple_spec_3 = cp_spec.add_traintuple(
+        dataset=dataset,
+        data_samples=[data_sample_3],
+        traintuple_specs=[traintuple_spec_2]
+    )
+    testtuple_spec_3 = cp_spec.add_testtuple(traintuple_spec_3)
+
+    # Submit compute plan and wait for it to complete
+    cp = session.add_compute_plan(cp_spec)
+
+    traintuples = [
+        session.get_traintuple(key).future().wait()
+        for key in cp.traintuple_keys
+    ]
+
+    testtuples = [
+        session.get_testtuple(key).future().wait()
+        for key in cp.testtuple_keys
+    ]
+
+    # All the train/test tuples should succeed
+    for t in traintuples + testtuples:
+        assert t.status == 'done'
+
+
+def test_compute_plan_single_session_failure(factory, session):
+    """In a compute plan with 3 traintuples, failing the root traintuple should also fail its descendents and the associated testtuples"""
```

Comment:
According to the new contributing guide for python code, we should avoid lines longer than 119 characters.